// ngx-bootstrap/accordion/bundles/ngx-bootstrap-accordion.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('ngx-bootstrap/utils'), require('@angular/common'), require('ngx-bootstrap/collapse')) :
    typeof define === 'function' && define.amd ? define('ngx-bootstrap/accordion', ['exports', '@angular/core', 'ngx-bootstrap/utils', '@angular/common', 'ngx-bootstrap/collapse'], factory) :
    (global = global || self, factory((global['ngx-bootstrap'] = global['ngx-bootstrap'] || {}, global['ngx-bootstrap'].accordion = {}), global.ng.core, global.utils, global.ng.common, global.collapse));
}(this, (function (exports, core, utils, common, collapse) { 'use strict';

    /**
     * Configuration service, provides default values for the AccordionComponent.
     */
    var AccordionConfig = /** @class */ (function () {
        function AccordionConfig() {
            /** Whether the other panels should be closed when a panel is opened */
            this.closeOthers = false;
            /** turn on/off animation */
            this.isAnimated = false;
        }
        /** @nocollapse */ AccordionConfig.ɵfac = function AccordionConfig_Factory(t) { return new (t || AccordionConfig)(); };
        /** @nocollapse */ AccordionConfig.ɵprov = core["ɵɵdefineInjectable"]({ token: AccordionConfig, factory: AccordionConfig.ɵfac });
        return AccordionConfig;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](AccordionConfig, [{
            type: core.Injectable
        }], null, null); })();

    var _c0 = ["*"];
    /** Displays collapsible content panels for presenting information in a limited amount of space. */
    var AccordionComponent = /** @class */ (function () {
        function AccordionComponent(config) {
            /** turn on/off animation */
            this.isAnimated = false;
            this.groups = [];
            Object.assign(this, config);
        }
        AccordionComponent.prototype.closeOtherPanels = function (openGroup) {
            if (!this.closeOthers) {
                return;
            }
            this.groups.forEach(function (group) {
                if (group !== openGroup) {
                    group.isOpen = false;
                }
            });
        };
        AccordionComponent.prototype.addGroup = function (group) {
            group.isAnimated = this.isAnimated;
            this.groups.push(group);
        };
        AccordionComponent.prototype.removeGroup = function (group) {
            var index = this.groups.indexOf(group);
            if (index !== -1) {
                this.groups.splice(index, 1);
            }
        };
        /** @nocollapse */ AccordionComponent.ɵfac = function AccordionComponent_Factory(t) { return new (t || AccordionComponent)(core["ɵɵdirectiveInject"](AccordionConfig)); };
        /** @nocollapse */ AccordionComponent.ɵcmp = core["ɵɵdefineComponent"]({ type: AccordionComponent, selectors: [["accordion"]], hostAttrs: ["role", "tablist", 1, "panel-group", 2, "display", "block"], hostVars: 1, hostBindings: function AccordionComponent_HostBindings(rf, ctx) { if (rf & 2) {
                core["ɵɵattribute"]("aria-multiselectable", ctx.closeOthers);
            } }, inputs: { isAnimated: "isAnimated", closeOthers: "closeOthers" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function AccordionComponent_Template(rf, ctx) { if (rf & 1) {
                core["ɵɵprojectionDef"]();
                core["ɵɵprojection"](0);
            } }, encapsulation: 2 });
        return AccordionComponent;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](AccordionComponent, [{
            type: core.Component,
            args: [{
                    selector: 'accordion',
                    template: "<ng-content></ng-content>",
                    host: {
                        '[attr.aria-multiselectable]': 'closeOthers',
                        role: 'tablist',
                        class: 'panel-group',
                        style: 'display: block'
                    }
                }]
        }], function () { return [{ type: AccordionConfig }]; }, { isAnimated: [{
                type: core.Input
            }], closeOthers: [{
                type: core.Input
            }] }); })();

    var _c0$1 = function (a0) { return { "text-muted": a0 }; };
    function AccordionPanelComponent_button_4_Template(rf, ctx) { if (rf & 1) {
        core["ɵɵelementStart"](0, "button", 7);
        core["ɵɵtext"](1);
        core["ɵɵelementEnd"]();
    } if (rf & 2) {
        var ctx_r1 = core["ɵɵnextContext"]();
        core["ɵɵproperty"]("ngClass", core["ɵɵpureFunction1"](2, _c0$1, ctx_r1.isDisabled));
        core["ɵɵadvance"](1);
        core["ɵɵtextInterpolate1"](" ", ctx_r1.heading, " ");
    } }
    var _c1 = [[["", "accordion-heading", ""]], "*"];
    var _c2 = ["[accordion-heading]", "*"];
    /**
     * ### Accordion heading
     * Instead of using `heading` attribute on the `accordion-group`, you can use
     * an `accordion-heading` attribute on `any` element inside of a group that
     * will be used as group's header template.
     */
    var AccordionPanelComponent = /** @class */ (function () {
        function AccordionPanelComponent(accordion) {
            /** turn on/off animation */
            this.isAnimated = false;
            /** Emits when the opened state changes */
            this.isOpenChange = new core.EventEmitter();
            this._isOpen = false;
            this.accordion = accordion;
        }
        Object.defineProperty(AccordionPanelComponent.prototype, "isOpen", {
            // Questionable, maybe .panel-open should be on child div.panel element?
            /** Is accordion group open or closed. This property supports two-way binding */
            get: function () {
                return this._isOpen;
            },
            set: function (value) {
                var _this = this;
                if (value !== this.isOpen) {
                    if (value) {
                        this.accordion.closeOtherPanels(this);
                    }
                    this._isOpen = value;
                    Promise.resolve(null).then(function () {
                        _this.isOpenChange.emit(value);
                    })
                        .catch(function (error) {
                        /* tslint:disable: no-console */
                        console.log(error);
                    });
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(AccordionPanelComponent.prototype, "isBs3", {
            get: function () {
                return utils.isBs3();
            },
            enumerable: true,
            configurable: true
        });
        AccordionPanelComponent.prototype.ngOnInit = function () {
            this.panelClass = this.panelClass || 'panel-default';
            this.accordion.addGroup(this);
        };
        AccordionPanelComponent.prototype.ngOnDestroy = function () {
            this.accordion.removeGroup(this);
        };
        AccordionPanelComponent.prototype.toggleOpen = function () {
            if (!this.isDisabled) {
                this.isOpen = !this.isOpen;
            }
        };
        /** @nocollapse */ AccordionPanelComponent.ɵfac = function AccordionPanelComponent_Factory(t) { return new (t || AccordionPanelComponent)(core["ɵɵdirectiveInject"](AccordionComponent)); };
        /** @nocollapse */ AccordionPanelComponent.ɵcmp = core["ɵɵdefineComponent"]({ type: AccordionPanelComponent, selectors: [["accordion-group"], ["accordion-panel"]], hostAttrs: [1, "panel", 2, "display", "block"], hostVars: 2, hostBindings: function AccordionPanelComponent_HostBindings(rf, ctx) { if (rf & 2) {
                core["ɵɵclassProp"]("panel-open", ctx.isOpen);
            } }, inputs: { heading: "heading", panelClass: "panelClass", isDisabled: "isDisabled", isOpen: "isOpen" }, outputs: { isOpenChange: "isOpenChange" }, ngContentSelectors: _c2, decls: 9, vars: 6, consts: [[1, "panel", "card", 3, "ngClass"], ["role", "tab", 1, "panel-heading", "card-header", 3, "ngClass", "click"], [1, "panel-title"], ["role", "button", 1, "accordion-toggle"], ["class", "btn btn-link", "type", "button", 3, "ngClass", 4, "ngIf"], ["role", "tabpanel", 1, "panel-collapse", "collapse", 3, "collapse", "isAnimated"], [1, "panel-body", "card-block", "card-body"], ["type", "button", 1, "btn", "btn-link", 3, "ngClass"]], template: function AccordionPanelComponent_Template(rf, ctx) { if (rf & 1) {
                core["ɵɵprojectionDef"](_c1);
                core["ɵɵelementStart"](0, "div", 0);
                core["ɵɵelementStart"](1, "div", 1);
                core["ɵɵlistener"]("click", function AccordionPanelComponent_Template_div_click_1_listener() { return ctx.toggleOpen(); });
                core["ɵɵelementStart"](2, "div", 2);
                core["ɵɵelementStart"](3, "div", 3);
                core["ɵɵtemplate"](4, AccordionPanelComponent_button_4_Template, 2, 4, "button", 4);
                core["ɵɵprojection"](5);
                core["ɵɵelementEnd"]();
                core["ɵɵelementEnd"]();
                core["ɵɵelementEnd"]();
                core["ɵɵelementStart"](6, "div", 5);
                core["ɵɵelementStart"](7, "div", 6);
                core["ɵɵprojection"](8, 1);
                core["ɵɵelementEnd"]();
                core["ɵɵelementEnd"]();
                core["ɵɵelementEnd"]();
            } if (rf & 2) {
                core["ɵɵproperty"]("ngClass", ctx.panelClass);
                core["ɵɵadvance"](1);
                core["ɵɵproperty"]("ngClass", ctx.isDisabled ? "panel-disabled" : "panel-enabled");
                core["ɵɵadvance"](2);
                core["ɵɵattribute"]("aria-expanded", ctx.isOpen);
                core["ɵɵadvance"](1);
                core["ɵɵproperty"]("ngIf", ctx.heading);
                core["ɵɵadvance"](2);
                core["ɵɵproperty"]("collapse", !ctx.isOpen)("isAnimated", ctx.isAnimated);
            } }, directives: [common.NgClass, common.NgIf, collapse.CollapseDirective], styles: ["[_nghost-%COMP%]   .card-header.panel-enabled[_ngcontent-%COMP%]{cursor:pointer}[_nghost-%COMP%]   .card-header.panel-disabled[_ngcontent-%COMP%]   .btn.btn-link[_ngcontent-%COMP%]{cursor:default;text-decoration:none}"] });
        return AccordionPanelComponent;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](AccordionPanelComponent, [{
            type: core.Component,
            args: [{
                    selector: 'accordion-group, accordion-panel',
                    templateUrl: './accordion-group.component.html',
                    host: {
                        class: 'panel',
                        style: 'display: block'
                    },
                    styleUrls: ['./accordion.scss']
                }]
        }], function () { return [{ type: AccordionComponent, decorators: [{
                    type: core.Inject,
                    args: [AccordionComponent]
                }] }]; }, { heading: [{
                type: core.Input
            }], panelClass: [{
                type: core.Input
            }], isDisabled: [{
                type: core.Input
            }], isOpenChange: [{
                type: core.Output
            }], isOpen: [{
                type: core.HostBinding,
                args: ['class.panel-open']
            }, {
                type: core.Input
            }] }); })();

    /* tslint:disable: max-classes-per-file */
    var AccordionModule = /** @class */ (function () {
        function AccordionModule() {
        }
        AccordionModule.forRoot = function () {
            return { ngModule: AccordionModule, providers: [AccordionConfig] };
        };
        /** @nocollapse */ AccordionModule.ɵmod = core["ɵɵdefineNgModule"]({ type: AccordionModule });
        /** @nocollapse */ AccordionModule.ɵinj = core["ɵɵdefineInjector"]({ factory: function AccordionModule_Factory(t) { return new (t || AccordionModule)(); }, imports: [[common.CommonModule, collapse.CollapseModule]] });
        return AccordionModule;
    }());
    (function () { (typeof ngJitMode === "undefined" || ngJitMode) && core["ɵɵsetNgModuleScope"](AccordionModule, { declarations: [AccordionComponent, AccordionPanelComponent], imports: [common.CommonModule, collapse.CollapseModule], exports: [AccordionComponent, AccordionPanelComponent] }); })();
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](AccordionModule, [{
            type: core.NgModule,
            args: [{
                    imports: [common.CommonModule, collapse.CollapseModule],
                    declarations: [AccordionComponent, AccordionPanelComponent],
                    exports: [AccordionComponent, AccordionPanelComponent]
                }]
        }], null, null); })();

    exports.AccordionComponent = AccordionComponent;
    exports.AccordionConfig = AccordionConfig;
    exports.AccordionModule = AccordionModule;
    exports.AccordionPanelComponent = AccordionPanelComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
