// ngx-bootstrap/pagination/bundles/ngx-bootstrap-pagination.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-bootstrap/pagination', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (global = global || self, factory((global['ngx-bootstrap'] = global['ngx-bootstrap'] || {}, global['ngx-bootstrap'].pagination = {}), global.ng.core, global.ng.forms, global.ng.common));
}(this, (function (exports, core, forms, common) { 'use strict';

    // todo: split
    /** Provides default values for Pagination and pager components */
    var PaginationConfig = /** @class */ (function () {
        function PaginationConfig() {
            this.main = {
                maxSize: void 0,
                itemsPerPage: 10,
                boundaryLinks: false,
                directionLinks: true,
                firstText: 'First',
                previousText: 'Previous',
                nextText: 'Next',
                lastText: 'Last',
                pageBtnClass: '',
                rotate: true
            };
            this.pager = {
                itemsPerPage: 15,
                previousText: '« Previous',
                nextText: 'Next »',
                pageBtnClass: '',
                align: true
            };
        }
        /** @nocollapse */ PaginationConfig.ɵfac = function PaginationConfig_Factory(t) { return new (t || PaginationConfig)(); };
        /** @nocollapse */ PaginationConfig.ɵprov = core["ɵɵdefineInjectable"]({ token: PaginationConfig, factory: PaginationConfig.ɵfac });
        return PaginationConfig;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](PaginationConfig, [{
            type: core.Injectable
        }], null, null); })();

    var _c0 = function (a0, a1) { return { "pull-left": a0, "float-left": a1 }; };
    var _c1 = function (a0, a1) { return { "pull-right": a0, "float-right": a1 }; };
    var PAGER_CONTROL_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        /* tslint:disable-next-line: no-use-before-declare */
        useExisting: core.forwardRef(function () { return PagerComponent; }),
        multi: true
    };
    var PagerComponent = /** @class */ (function () {
        function PagerComponent(elementRef, paginationConfig, changeDetection) {
            this.elementRef = elementRef;
            this.changeDetection = changeDetection;
            /** fired when total pages count changes, $event:number equals to total pages count */
            this.numPages = new core.EventEmitter();
            /** fired when page was changed, $event:{page, itemsPerPage} equals to
             * object with current page index and number of items per page
             */
            this.pageChanged = new core.EventEmitter();
            this.onChange = Function.prototype;
            this.onTouched = Function.prototype;
            this.inited = false;
            this._page = 1;
            this.elementRef = elementRef;
            if (!this.config) {
                this.configureOptions(Object.assign({}, paginationConfig.main, paginationConfig.pager));
            }
        }
        Object.defineProperty(PagerComponent.prototype, "itemsPerPage", {
            /** maximum number of items per page. If value less than 1 will display all items on one page */
            get: function () {
                return this._itemsPerPage;
            },
            set: function (v) {
                this._itemsPerPage = v;
                this.totalPages = this.calculateTotalPages();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PagerComponent.prototype, "totalItems", {
            /** total number of items in all pages */
            get: function () {
                return this._totalItems;
            },
            set: function (v) {
                this._totalItems = v;
                this.totalPages = this.calculateTotalPages();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PagerComponent.prototype, "totalPages", {
            get: function () {
                return this._totalPages;
            },
            set: function (v) {
                this._totalPages = v;
                this.numPages.emit(v);
                if (this.inited) {
                    this.selectPage(this.page);
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PagerComponent.prototype, "page", {
            get: function () {
                return this._page;
            },
            set: function (value) {
                var _previous = this._page;
                this._page = value > this.totalPages ? this.totalPages : value || 1;
                this.changeDetection.markForCheck();
                if (_previous === this._page || typeof _previous === 'undefined') {
                    return;
                }
                this.pageChanged.emit({
                    page: this._page,
                    itemsPerPage: this.itemsPerPage
                });
            },
            enumerable: true,
            configurable: true
        });
        PagerComponent.prototype.configureOptions = function (config) {
            this.config = Object.assign({}, config);
        };
        PagerComponent.prototype.ngOnInit = function () {
            if (typeof window !== 'undefined') {
                this.classMap = this.elementRef.nativeElement.getAttribute('class') || '';
            }
            // watch for maxSize
            this.maxSize =
                typeof this.maxSize !== 'undefined' ? this.maxSize : this.config.maxSize;
            this.rotate =
                typeof this.rotate !== 'undefined' ? this.rotate : this.config.rotate;
            this.boundaryLinks =
                typeof this.boundaryLinks !== 'undefined'
                    ? this.boundaryLinks
                    : this.config.boundaryLinks;
            this.directionLinks =
                typeof this.directionLinks !== 'undefined'
                    ? this.directionLinks
                    : this.config.directionLinks;
            this.pageBtnClass =
                typeof this.pageBtnClass !== 'undefined'
                    ? this.pageBtnClass
                    : this.config.pageBtnClass;
            // base class
            this.itemsPerPage =
                typeof this.itemsPerPage !== 'undefined'
                    ? this.itemsPerPage
                    : this.config.itemsPerPage;
            this.totalPages = this.calculateTotalPages();
            // this class
            this.pages = this.getPages(this.page, this.totalPages);
            this.inited = true;
        };
        PagerComponent.prototype.writeValue = function (value) {
            this.page = value;
            this.pages = this.getPages(this.page, this.totalPages);
        };
        PagerComponent.prototype.getText = function (key) {
            // tslint:disable-next-line:no-any
            return this[key + "Text"] || this.config[key + "Text"];
        };
        PagerComponent.prototype.noPrevious = function () {
            return this.page === 1;
        };
        PagerComponent.prototype.noNext = function () {
            return this.page === this.totalPages;
        };
        PagerComponent.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        PagerComponent.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        PagerComponent.prototype.selectPage = function (page, event) {
            if (event) {
                event.preventDefault();
            }
            if (!this.disabled) {
                if (event && event.target) {
                    // tslint:disable-next-line:no-any
                    var target = event.target;
                    target.blur();
                }
                this.writeValue(page);
                this.onChange(this.page);
            }
        };
        // Create page object used in template
        PagerComponent.prototype.makePage = function (num, text, active) {
            return { text: text, number: num, active: active };
        };
        PagerComponent.prototype.getPages = function (currentPage, totalPages) {
            var pages = [];
            // Default page limits
            var startPage = 1;
            var endPage = totalPages;
            var isMaxSized = typeof this.maxSize !== 'undefined' && this.maxSize < totalPages;
            // recompute if maxSize
            if (isMaxSized) {
                if (this.rotate) {
                    // Current page is displayed in the middle of the visible ones
                    startPage = Math.max(currentPage - Math.floor(this.maxSize / 2), 1);
                    endPage = startPage + this.maxSize - 1;
                    // Adjust if limit is exceeded
                    if (endPage > totalPages) {
                        endPage = totalPages;
                        startPage = endPage - this.maxSize + 1;
                    }
                }
                else {
                    // Visible pages are paginated with maxSize
                    startPage =
                        (Math.ceil(currentPage / this.maxSize) - 1) * this.maxSize + 1;
                    // Adjust last page if limit is exceeded
                    endPage = Math.min(startPage + this.maxSize - 1, totalPages);
                }
            }
            // Add page number links
            for (var num = startPage; num <= endPage; num++) {
                var page = this.makePage(num, num.toString(), num === currentPage);
                pages.push(page);
            }
            // Add links to move between page sets
            if (isMaxSized && !this.rotate) {
                if (startPage > 1) {
                    var previousPageSet = this.makePage(startPage - 1, '...', false);
                    pages.unshift(previousPageSet);
                }
                if (endPage < totalPages) {
                    var nextPageSet = this.makePage(endPage + 1, '...', false);
                    pages.push(nextPageSet);
                }
            }
            return pages;
        };
        // base class
        PagerComponent.prototype.calculateTotalPages = function () {
            var totalPages = this.itemsPerPage < 1
                ? 1
                : Math.ceil(this.totalItems / this.itemsPerPage);
            return Math.max(totalPages || 0, 1);
        };
        /** @nocollapse */ PagerComponent.ɵfac = function PagerComponent_Factory(t) { return new (t || PagerComponent)(core["ɵɵdirectiveInject"](core.ElementRef), core["ɵɵdirectiveInject"](PaginationConfig), core["ɵɵdirectiveInject"](core.ChangeDetectorRef)); };
        /** @nocollapse */ PagerComponent.ɵcmp = core["ɵɵdefineComponent"]({ type: PagerComponent, selectors: [["pager"]], inputs: { align: "align", maxSize: "maxSize", boundaryLinks: "boundaryLinks", directionLinks: "directionLinks", firstText: "firstText", previousText: "previousText", nextText: "nextText", lastText: "lastText", rotate: "rotate", pageBtnClass: "pageBtnClass", disabled: "disabled", itemsPerPage: "itemsPerPage", totalItems: "totalItems" }, outputs: { numPages: "numPages", pageChanged: "pageChanged" }, features: [core["ɵɵProvidersFeature"]([PAGER_CONTROL_VALUE_ACCESSOR])], decls: 7, vars: 24, consts: [[1, "pager"], [3, "ngClass"], ["href", "", 3, "click"]], template: function PagerComponent_Template(rf, ctx) { if (rf & 1) {
                core["ɵɵelementStart"](0, "ul", 0);
                core["ɵɵelementStart"](1, "li", 1);
                core["ɵɵelementStart"](2, "a", 2);
                core["ɵɵlistener"]("click", function PagerComponent_Template_a_click_2_listener($event) { return ctx.selectPage(ctx.page - 1, $event); });
                core["ɵɵtext"](3);
                core["ɵɵelementEnd"]();
                core["ɵɵelementEnd"]();
                core["ɵɵelementStart"](4, "li", 1);
                core["ɵɵelementStart"](5, "a", 2);
                core["ɵɵlistener"]("click", function PagerComponent_Template_a_click_5_listener($event) { return ctx.selectPage(ctx.page + 1, $event); });
                core["ɵɵtext"](6);
                core["ɵɵelementEnd"]();
                core["ɵɵelementEnd"]();
                core["ɵɵelementEnd"]();
            } if (rf & 2) {
                core["ɵɵadvance"](1);
                core["ɵɵclassMap"](ctx.pageBtnClass);
                core["ɵɵclassProp"]("disabled", ctx.noPrevious())("previous", ctx.align);
                core["ɵɵproperty"]("ngClass", core["ɵɵpureFunction2"](18, _c0, ctx.align, ctx.align));
                core["ɵɵadvance"](2);
                core["ɵɵtextInterpolate"](ctx.getText("previous"));
                core["ɵɵadvance"](1);
                core["ɵɵclassMap"](ctx.pageBtnClass);
                core["ɵɵclassProp"]("disabled", ctx.noNext())("next", ctx.align);
                core["ɵɵproperty"]("ngClass", core["ɵɵpureFunction2"](21, _c1, ctx.align, ctx.align));
                core["ɵɵadvance"](2);
                core["ɵɵtextInterpolate"](ctx.getText("next"));
            } }, directives: [common.NgClass], encapsulation: 2 });
        return PagerComponent;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](PagerComponent, [{
            type: core.Component,
            args: [{
                    selector: 'pager',
                    templateUrl: './pager.component.html',
                    providers: [PAGER_CONTROL_VALUE_ACCESSOR]
                }]
        }], function () { return [{ type: core.ElementRef }, { type: PaginationConfig }, { type: core.ChangeDetectorRef }]; }, { align: [{
                type: core.Input
            }], maxSize: [{
                type: core.Input
            }], boundaryLinks: [{
                type: core.Input
            }], directionLinks: [{
                type: core.Input
            }], firstText: [{
                type: core.Input
            }], previousText: [{
                type: core.Input
            }], nextText: [{
                type: core.Input
            }], lastText: [{
                type: core.Input
            }], rotate: [{
                type: core.Input
            }], pageBtnClass: [{
                type: core.Input
            }], disabled: [{
                type: core.Input
            }], numPages: [{
                type: core.Output
            }], pageChanged: [{
                type: core.Output
            }], itemsPerPage: [{
                type: core.Input
            }], totalItems: [{
                type: core.Input
            }] }); })();

    var _c0$1 = function (a0, a1) { return { disabled: a0, currentPage: a1 }; };
    function PaginationComponent_li_1_Template(rf, ctx) { if (rf & 1) {
        var _r43 = core["ɵɵgetCurrentView"]();
        core["ɵɵelementStart"](0, "li", 11);
        core["ɵɵelementStart"](1, "a", 12);
        core["ɵɵlistener"]("click", function PaginationComponent_li_1_Template_a_click_1_listener($event) { core["ɵɵrestoreView"](_r43); var ctx_r42 = core["ɵɵnextContext"](); return ctx_r42.selectPage(1, $event); });
        core["ɵɵelementContainer"](2, 13);
        core["ɵɵelementEnd"]();
        core["ɵɵelementEnd"]();
    } if (rf & 2) {
        var ctx_r27 = core["ɵɵnextContext"]();
        var _r38 = core["ɵɵreference"](13);
        core["ɵɵclassProp"]("disabled", ctx_r27.noPrevious() || ctx_r27.disabled);
        core["ɵɵadvance"](2);
        core["ɵɵproperty"]("ngTemplateOutlet", ctx_r27.customFirstTemplate || _r38)("ngTemplateOutletContext", core["ɵɵpureFunction2"](4, _c0$1, ctx_r27.noPrevious() || ctx_r27.disabled, ctx_r27.page));
    } }
    function PaginationComponent_li_2_Template(rf, ctx) { if (rf & 1) {
        var _r45 = core["ɵɵgetCurrentView"]();
        core["ɵɵelementStart"](0, "li", 14);
        core["ɵɵelementStart"](1, "a", 12);
        core["ɵɵlistener"]("click", function PaginationComponent_li_2_Template_a_click_1_listener($event) { core["ɵɵrestoreView"](_r45); var ctx_r44 = core["ɵɵnextContext"](); return ctx_r44.selectPage(ctx_r44.page - 1, $event); });
        core["ɵɵelementContainer"](2, 13);
        core["ɵɵelementEnd"]();
        core["ɵɵelementEnd"]();
    } if (rf & 2) {
        var ctx_r28 = core["ɵɵnextContext"]();
        var _r36 = core["ɵɵreference"](11);
        core["ɵɵclassProp"]("disabled", ctx_r28.noPrevious() || ctx_r28.disabled);
        core["ɵɵadvance"](2);
        core["ɵɵproperty"]("ngTemplateOutlet", ctx_r28.customPreviousTemplate || _r36)("ngTemplateOutletContext", core["ɵɵpureFunction2"](4, _c0$1, ctx_r28.noPrevious() || ctx_r28.disabled, ctx_r28.page));
    } }
    var _c1$1 = function (a0, a1, a2) { return { disabled: a0, $implicit: a1, currentPage: a2 }; };
    function PaginationComponent_li_3_Template(rf, ctx) { if (rf & 1) {
        var _r48 = core["ɵɵgetCurrentView"]();
        core["ɵɵelementStart"](0, "li", 15);
        core["ɵɵelementStart"](1, "a", 12);
        core["ɵɵlistener"]("click", function PaginationComponent_li_3_Template_a_click_1_listener($event) { core["ɵɵrestoreView"](_r48); var pg_r46 = ctx.$implicit; var ctx_r47 = core["ɵɵnextContext"](); return ctx_r47.selectPage(pg_r46.number, $event); });
        core["ɵɵelementContainer"](2, 13);
        core["ɵɵelementEnd"]();
        core["ɵɵelementEnd"]();
    } if (rf & 2) {
        var pg_r46 = ctx.$implicit;
        var ctx_r29 = core["ɵɵnextContext"]();
        var _r32 = core["ɵɵreference"](7);
        core["ɵɵclassProp"]("active", pg_r46.active)("disabled", ctx_r29.disabled && !pg_r46.active);
        core["ɵɵadvance"](2);
        core["ɵɵproperty"]("ngTemplateOutlet", ctx_r29.customPageTemplate || _r32)("ngTemplateOutletContext", core["ɵɵpureFunction3"](6, _c1$1, ctx_r29.disabled, pg_r46, ctx_r29.page));
    } }
    function PaginationComponent_li_4_Template(rf, ctx) { if (rf & 1) {
        var _r50 = core["ɵɵgetCurrentView"]();
        core["ɵɵelementStart"](0, "li", 16);
        core["ɵɵelementStart"](1, "a", 12);
        core["ɵɵlistener"]("click", function PaginationComponent_li_4_Template_a_click_1_listener($event) { core["ɵɵrestoreView"](_r50); var ctx_r49 = core["ɵɵnextContext"](); return ctx_r49.selectPage(ctx_r49.page + 1, $event); });
        core["ɵɵelementContainer"](2, 13);
        core["ɵɵelementEnd"]();
        core["ɵɵelementEnd"]();
    } if (rf & 2) {
        var ctx_r30 = core["ɵɵnextContext"]();
        var _r34 = core["ɵɵreference"](9);
        core["ɵɵclassProp"]("disabled", ctx_r30.noNext() || ctx_r30.disabled);
        core["ɵɵadvance"](2);
        core["ɵɵproperty"]("ngTemplateOutlet", ctx_r30.customNextTemplate || _r34)("ngTemplateOutletContext", core["ɵɵpureFunction2"](4, _c0$1, ctx_r30.noNext() || ctx_r30.disabled, ctx_r30.page));
    } }
    function PaginationComponent_li_5_Template(rf, ctx) { if (rf & 1) {
        var _r52 = core["ɵɵgetCurrentView"]();
        core["ɵɵelementStart"](0, "li", 17);
        core["ɵɵelementStart"](1, "a", 12);
        core["ɵɵlistener"]("click", function PaginationComponent_li_5_Template_a_click_1_listener($event) { core["ɵɵrestoreView"](_r52); var ctx_r51 = core["ɵɵnextContext"](); return ctx_r51.selectPage(ctx_r51.totalPages, $event); });
        core["ɵɵelementContainer"](2, 13);
        core["ɵɵelementEnd"]();
        core["ɵɵelementEnd"]();
    } if (rf & 2) {
        var ctx_r31 = core["ɵɵnextContext"]();
        var _r40 = core["ɵɵreference"](15);
        core["ɵɵclassProp"]("disabled", ctx_r31.noNext() || ctx_r31.disabled);
        core["ɵɵadvance"](2);
        core["ɵɵproperty"]("ngTemplateOutlet", ctx_r31.customLastTemplate || _r40)("ngTemplateOutletContext", core["ɵɵpureFunction2"](4, _c0$1, ctx_r31.noNext() || ctx_r31.disabled, ctx_r31.page));
    } }
    function PaginationComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
        core["ɵɵtext"](0);
    } if (rf & 2) {
        var page_r53 = ctx.$implicit;
        core["ɵɵtextInterpolate"](page_r53.text);
    } }
    function PaginationComponent_ng_template_8_Template(rf, ctx) { if (rf & 1) {
        core["ɵɵtext"](0);
    } if (rf & 2) {
        var ctx_r35 = core["ɵɵnextContext"]();
        core["ɵɵtextInterpolate"](ctx_r35.getText("next"));
    } }
    function PaginationComponent_ng_template_10_Template(rf, ctx) { if (rf & 1) {
        core["ɵɵtext"](0);
    } if (rf & 2) {
        var ctx_r37 = core["ɵɵnextContext"]();
        core["ɵɵtextInterpolate"](ctx_r37.getText("previous"));
    } }
    function PaginationComponent_ng_template_12_Template(rf, ctx) { if (rf & 1) {
        core["ɵɵtext"](0);
    } if (rf & 2) {
        var ctx_r39 = core["ɵɵnextContext"]();
        core["ɵɵtextInterpolate"](ctx_r39.getText("first"));
    } }
    function PaginationComponent_ng_template_14_Template(rf, ctx) { if (rf & 1) {
        core["ɵɵtext"](0);
    } if (rf & 2) {
        var ctx_r41 = core["ɵɵnextContext"]();
        core["ɵɵtextInterpolate"](ctx_r41.getText("last"));
    } }
    var PAGINATION_CONTROL_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        /* tslint:disable-next-line: no-use-before-declare */
        useExisting: core.forwardRef(function () { return PaginationComponent; }),
        multi: true
    };
    var PaginationComponent = /** @class */ (function () {
        function PaginationComponent(elementRef, paginationConfig, changeDetection) {
            this.elementRef = elementRef;
            this.changeDetection = changeDetection;
            /** fired when total pages count changes, $event:number equals to total pages count */
            this.numPages = new core.EventEmitter();
            /** fired when page was changed, $event:{page, itemsPerPage} equals to object
             * with current page index and number of items per page
             */
            this.pageChanged = new core.EventEmitter();
            this.onChange = Function.prototype;
            this.onTouched = Function.prototype;
            this.inited = false;
            this._page = 1;
            this.elementRef = elementRef;
            if (!this.config) {
                this.configureOptions(paginationConfig.main);
            }
        }
        Object.defineProperty(PaginationComponent.prototype, "itemsPerPage", {
            /** maximum number of items per page. If value less than 1 will display all items on one page */
            get: function () {
                return this._itemsPerPage;
            },
            set: function (v) {
                this._itemsPerPage = v;
                this.totalPages = this.calculateTotalPages();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PaginationComponent.prototype, "totalItems", {
            /** total number of items in all pages */
            get: function () {
                return this._totalItems;
            },
            set: function (v) {
                this._totalItems = v;
                this.totalPages = this.calculateTotalPages();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PaginationComponent.prototype, "totalPages", {
            get: function () {
                return this._totalPages;
            },
            set: function (v) {
                this._totalPages = v;
                this.numPages.emit(v);
                if (this.inited) {
                    this.selectPage(this.page);
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PaginationComponent.prototype, "page", {
            get: function () {
                return this._page;
            },
            set: function (value) {
                var _previous = this._page;
                this._page = value > this.totalPages ? this.totalPages : value || 1;
                this.changeDetection.markForCheck();
                if (_previous === this._page || typeof _previous === 'undefined') {
                    return;
                }
                this.pageChanged.emit({
                    page: this._page,
                    itemsPerPage: this.itemsPerPage
                });
            },
            enumerable: true,
            configurable: true
        });
        PaginationComponent.prototype.configureOptions = function (config) {
            this.config = Object.assign({}, config);
        };
        PaginationComponent.prototype.ngOnInit = function () {
            if (typeof window !== 'undefined') {
                this.classMap = this.elementRef.nativeElement.getAttribute('class') || '';
            }
            // watch for maxSize
            this.maxSize =
                typeof this.maxSize !== 'undefined' ? this.maxSize : this.config.maxSize;
            this.rotate =
                typeof this.rotate !== 'undefined' ? this.rotate : this.config.rotate;
            this.boundaryLinks =
                typeof this.boundaryLinks !== 'undefined'
                    ? this.boundaryLinks
                    : this.config.boundaryLinks;
            this.directionLinks =
                typeof this.directionLinks !== 'undefined'
                    ? this.directionLinks
                    : this.config.directionLinks;
            this.pageBtnClass =
                typeof this.pageBtnClass !== 'undefined'
                    ? this.pageBtnClass
                    : this.config.pageBtnClass;
            // base class
            this.itemsPerPage =
                typeof this.itemsPerPage !== 'undefined'
                    ? this.itemsPerPage
                    : this.config.itemsPerPage;
            this.totalPages = this.calculateTotalPages();
            // this class
            this.pages = this.getPages(this.page, this.totalPages);
            this.inited = true;
        };
        PaginationComponent.prototype.writeValue = function (value) {
            this.page = value;
            this.pages = this.getPages(this.page, this.totalPages);
        };
        PaginationComponent.prototype.getText = function (key) {
            // tslint:disable-next-line:no-any
            return this[key + "Text"] || this.config[key + "Text"];
        };
        PaginationComponent.prototype.noPrevious = function () {
            return this.page === 1;
        };
        PaginationComponent.prototype.noNext = function () {
            return this.page === this.totalPages;
        };
        PaginationComponent.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        PaginationComponent.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        PaginationComponent.prototype.selectPage = function (page, event) {
            if (event) {
                event.preventDefault();
            }
            if (!this.disabled) {
                if (event && event.target) {
                    // tslint:disable-next-line:no-any
                    var target = event.target;
                    target.blur();
                }
                this.writeValue(page);
                this.onChange(this.page);
            }
        };
        // Create page object used in template
        PaginationComponent.prototype.makePage = function (num, text, active) {
            return { text: text, number: num, active: active };
        };
        PaginationComponent.prototype.getPages = function (currentPage, totalPages) {
            var pages = [];
            // Default page limits
            var startPage = 1;
            var endPage = totalPages;
            var isMaxSized = typeof this.maxSize !== 'undefined' && this.maxSize < totalPages;
            // recompute if maxSize
            if (isMaxSized) {
                if (this.rotate) {
                    // Current page is displayed in the middle of the visible ones
                    startPage = Math.max(currentPage - Math.floor(this.maxSize / 2), 1);
                    endPage = startPage + this.maxSize - 1;
                    // Adjust if limit is exceeded
                    if (endPage > totalPages) {
                        endPage = totalPages;
                        startPage = endPage - this.maxSize + 1;
                    }
                }
                else {
                    // Visible pages are paginated with maxSize
                    startPage =
                        (Math.ceil(currentPage / this.maxSize) - 1) * this.maxSize + 1;
                    // Adjust last page if limit is exceeded
                    endPage = Math.min(startPage + this.maxSize - 1, totalPages);
                }
            }
            // Add page number links
            for (var num = startPage; num <= endPage; num++) {
                var page = this.makePage(num, num.toString(), num === currentPage);
                pages.push(page);
            }
            // Add links to move between page sets
            if (isMaxSized && !this.rotate) {
                if (startPage > 1) {
                    var previousPageSet = this.makePage(startPage - 1, '...', false);
                    pages.unshift(previousPageSet);
                }
                if (endPage < totalPages) {
                    var nextPageSet = this.makePage(endPage + 1, '...', false);
                    pages.push(nextPageSet);
                }
            }
            return pages;
        };
        // base class
        PaginationComponent.prototype.calculateTotalPages = function () {
            var totalPages = this.itemsPerPage < 1
                ? 1
                : Math.ceil(this.totalItems / this.itemsPerPage);
            return Math.max(totalPages || 0, 1);
        };
        /** @nocollapse */ PaginationComponent.ɵfac = function PaginationComponent_Factory(t) { return new (t || PaginationComponent)(core["ɵɵdirectiveInject"](core.ElementRef), core["ɵɵdirectiveInject"](PaginationConfig), core["ɵɵdirectiveInject"](core.ChangeDetectorRef)); };
        /** @nocollapse */ PaginationComponent.ɵcmp = core["ɵɵdefineComponent"]({ type: PaginationComponent, selectors: [["pagination"]], inputs: { align: "align", maxSize: "maxSize", boundaryLinks: "boundaryLinks", directionLinks: "directionLinks", firstText: "firstText", previousText: "previousText", nextText: "nextText", lastText: "lastText", rotate: "rotate", pageBtnClass: "pageBtnClass", disabled: "disabled", customPageTemplate: "customPageTemplate", customNextTemplate: "customNextTemplate", customPreviousTemplate: "customPreviousTemplate", customFirstTemplate: "customFirstTemplate", customLastTemplate: "customLastTemplate", itemsPerPage: "itemsPerPage", totalItems: "totalItems" }, outputs: { numPages: "numPages", pageChanged: "pageChanged" }, features: [core["ɵɵProvidersFeature"]([PAGINATION_CONTROL_VALUE_ACCESSOR])], decls: 16, vars: 6, consts: [[1, "pagination", 3, "ngClass"], ["class", "pagination-first page-item", 3, "disabled", 4, "ngIf"], ["class", "pagination-prev page-item", 3, "disabled", 4, "ngIf"], ["class", "pagination-page page-item", 3, "active", "disabled", 4, "ngFor", "ngForOf"], ["class", "pagination-next page-item", 3, "disabled", 4, "ngIf"], ["class", "pagination-last page-item", 3, "disabled", 4, "ngIf"], ["defaultPageTemplate", ""], ["defaultNextTemplate", ""], ["defaultPreviousTemplate", ""], ["defaultFirstTemplate", ""], ["defaultLastTemplate", ""], [1, "pagination-first", "page-item"], ["href", "", 1, "page-link", 3, "click"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "pagination-prev", "page-item"], [1, "pagination-page", "page-item"], [1, "pagination-next", "page-item"], [1, "pagination-last", "page-item"]], template: function PaginationComponent_Template(rf, ctx) { if (rf & 1) {
                core["ɵɵelementStart"](0, "ul", 0);
                core["ɵɵtemplate"](1, PaginationComponent_li_1_Template, 3, 7, "li", 1);
                core["ɵɵtemplate"](2, PaginationComponent_li_2_Template, 3, 7, "li", 2);
                core["ɵɵtemplate"](3, PaginationComponent_li_3_Template, 3, 10, "li", 3);
                core["ɵɵtemplate"](4, PaginationComponent_li_4_Template, 3, 7, "li", 4);
                core["ɵɵtemplate"](5, PaginationComponent_li_5_Template, 3, 7, "li", 5);
                core["ɵɵelementEnd"]();
                core["ɵɵtemplate"](6, PaginationComponent_ng_template_6_Template, 1, 1, "ng-template", null, 6, core["ɵɵtemplateRefExtractor"]);
                core["ɵɵtemplate"](8, PaginationComponent_ng_template_8_Template, 1, 1, "ng-template", null, 7, core["ɵɵtemplateRefExtractor"]);
                core["ɵɵtemplate"](10, PaginationComponent_ng_template_10_Template, 1, 1, "ng-template", null, 8, core["ɵɵtemplateRefExtractor"]);
                core["ɵɵtemplate"](12, PaginationComponent_ng_template_12_Template, 1, 1, "ng-template", null, 9, core["ɵɵtemplateRefExtractor"]);
                core["ɵɵtemplate"](14, PaginationComponent_ng_template_14_Template, 1, 1, "ng-template", null, 10, core["ɵɵtemplateRefExtractor"]);
            } if (rf & 2) {
                core["ɵɵproperty"]("ngClass", ctx.classMap);
                core["ɵɵadvance"](1);
                core["ɵɵproperty"]("ngIf", ctx.boundaryLinks);
                core["ɵɵadvance"](1);
                core["ɵɵproperty"]("ngIf", ctx.directionLinks);
                core["ɵɵadvance"](1);
                core["ɵɵproperty"]("ngForOf", ctx.pages);
                core["ɵɵadvance"](1);
                core["ɵɵproperty"]("ngIf", ctx.directionLinks);
                core["ɵɵadvance"](1);
                core["ɵɵproperty"]("ngIf", ctx.boundaryLinks);
            } }, directives: [common.NgClass, common.NgIf, common.NgForOf, common.NgTemplateOutlet], encapsulation: 2 });
        return PaginationComponent;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](PaginationComponent, [{
            type: core.Component,
            args: [{
                    selector: 'pagination',
                    templateUrl: './pagination.component.html',
                    providers: [PAGINATION_CONTROL_VALUE_ACCESSOR]
                }]
        }], function () { return [{ type: core.ElementRef }, { type: PaginationConfig }, { type: core.ChangeDetectorRef }]; }, { align: [{
                type: core.Input
            }], maxSize: [{
                type: core.Input
            }], boundaryLinks: [{
                type: core.Input
            }], directionLinks: [{
                type: core.Input
            }], firstText: [{
                type: core.Input
            }], previousText: [{
                type: core.Input
            }], nextText: [{
                type: core.Input
            }], lastText: [{
                type: core.Input
            }], rotate: [{
                type: core.Input
            }], pageBtnClass: [{
                type: core.Input
            }], disabled: [{
                type: core.Input
            }], customPageTemplate: [{
                type: core.Input
            }], customNextTemplate: [{
                type: core.Input
            }], customPreviousTemplate: [{
                type: core.Input
            }], customFirstTemplate: [{
                type: core.Input
            }], customLastTemplate: [{
                type: core.Input
            }], numPages: [{
                type: core.Output
            }], pageChanged: [{
                type: core.Output
            }], itemsPerPage: [{
                type: core.Input
            }], totalItems: [{
                type: core.Input
            }] }); })();

    var PaginationModule = /** @class */ (function () {
        function PaginationModule() {
        }
        PaginationModule.forRoot = function () {
            return { ngModule: PaginationModule, providers: [PaginationConfig] };
        };
        /** @nocollapse */ PaginationModule.ɵmod = core["ɵɵdefineNgModule"]({ type: PaginationModule });
        /** @nocollapse */ PaginationModule.ɵinj = core["ɵɵdefineInjector"]({ factory: function PaginationModule_Factory(t) { return new (t || PaginationModule)(); }, imports: [[common.CommonModule]] });
        return PaginationModule;
    }());
    (function () { (typeof ngJitMode === "undefined" || ngJitMode) && core["ɵɵsetNgModuleScope"](PaginationModule, { declarations: [PagerComponent, PaginationComponent], imports: [common.CommonModule], exports: [PagerComponent, PaginationComponent] }); })();
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](PaginationModule, [{
            type: core.NgModule,
            args: [{
                    imports: [common.CommonModule],
                    declarations: [PagerComponent, PaginationComponent],
                    exports: [PagerComponent, PaginationComponent]
                }]
        }], null, null); })();

    exports.PagerComponent = PagerComponent;
    exports.PaginationComponent = PaginationComponent;
    exports.PaginationConfig = PaginationConfig;
    exports.PaginationModule = PaginationModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
