// ngx-bootstrap/rating/bundles/ngx-bootstrap-rating.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-bootstrap/rating', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (global = global || self, factory((global['ngx-bootstrap'] = global['ngx-bootstrap'] || {}, global['ngx-bootstrap'].rating = {}), global.ng.core, global.ng.forms, global.ng.common));
}(this, (function (exports, core, forms, common) { 'use strict';

    /** Default values provider for rating */
    var RatingConfig = /** @class */ (function () {
        function RatingConfig() {
            /** aria label for rating */
            this.ariaLabel = 'rating';
        }
        /** @nocollapse */ RatingConfig.ɵfac = function RatingConfig_Factory(t) { return new (t || RatingConfig)(); };
        /** @nocollapse */ RatingConfig.ɵprov = core["ɵɵdefineInjectable"]({ token: RatingConfig, factory: RatingConfig.ɵfac });
        return RatingConfig;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](RatingConfig, [{
            type: core.Injectable
        }], null, null); })();

    function RatingComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
        core["ɵɵtext"](0);
    } if (rf & 2) {
        var value_r14 = ctx.value;
        var index_r15 = ctx.index;
        core["ɵɵtextInterpolate"](index_r15 < value_r14 ? "\u2605" : "\u2606");
    } }
    function RatingComponent_ng_template_3_ng_template_3_Template(rf, ctx) { }
    var _c0 = function (a0, a1) { return { index: a0, value: a1 }; };
    function RatingComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
        var _r20 = core["ɵɵgetCurrentView"]();
        core["ɵɵelementStart"](0, "span", 3);
        core["ɵɵtext"](1);
        core["ɵɵelementEnd"]();
        core["ɵɵelementStart"](2, "span", 4);
        core["ɵɵlistener"]("mouseenter", function RatingComponent_ng_template_3_Template_span_mouseenter_2_listener() { core["ɵɵrestoreView"](_r20); var index_r17 = ctx.index; var ctx_r19 = core["ɵɵnextContext"](); return ctx_r19.enter(index_r17 + 1); })("click", function RatingComponent_ng_template_3_Template_span_click_2_listener() { core["ɵɵrestoreView"](_r20); var index_r17 = ctx.index; var ctx_r21 = core["ɵɵnextContext"](); return ctx_r21.rate(index_r17 + 1); });
        core["ɵɵtemplate"](3, RatingComponent_ng_template_3_ng_template_3_Template, 0, 0, "ng-template", 5);
        core["ɵɵelementEnd"]();
    } if (rf & 2) {
        var r_r16 = ctx.$implicit;
        var index_r17 = ctx.index;
        var ctx_r13 = core["ɵɵnextContext"]();
        var _r11 = core["ɵɵreference"](2);
        core["ɵɵadvance"](1);
        core["ɵɵtextInterpolate1"]("(", index_r17 < ctx_r13.value ? "*" : " ", ")");
        core["ɵɵadvance"](1);
        core["ɵɵstyleProp"]("cursor", ctx_r13.readonly ? "default" : "pointer");
        core["ɵɵclassProp"]("active", index_r17 < ctx_r13.value);
        core["ɵɵproperty"]("title", r_r16.title);
        core["ɵɵadvance"](1);
        core["ɵɵproperty"]("ngTemplateOutlet", ctx_r13.customTemplate || _r11)("ngTemplateOutletContext", core["ɵɵpureFunction2"](8, _c0, index_r17, ctx_r13.value));
    } }
    var RATING_CONTROL_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        /* tslint:disable-next-line: no-use-before-declare */
        useExisting: core.forwardRef(function () { return RatingComponent; }),
        multi: true
    };
    var RatingComponent = /** @class */ (function () {
        function RatingComponent(changeDetection, config) {
            this.changeDetection = changeDetection;
            /** number of icons */
            this.max = 5;
            /** fired when icon selected, $event:number equals to selected rating */
            this.onHover = new core.EventEmitter();
            /** fired when icon selected, $event:number equals to previous rating value */
            this.onLeave = new core.EventEmitter();
            // tslint:disable-next-line:no-any
            this.onChange = Function.prototype;
            // tslint:disable-next-line:no-any
            this.onTouched = Function.prototype;
            Object.assign(this, config);
        }
        RatingComponent.prototype.onKeydown = function (event) {
            /* tslint:disable-next-line: deprecation */
            if ([37, 38, 39, 40].indexOf(event.which) === -1) {
                return;
            }
            event.preventDefault();
            event.stopPropagation();
            /* tslint:disable-next-line: deprecation */
            var sign = event.which === 38 || event.which === 39 ? 1 : -1;
            this.rate(this.value + sign);
        };
        RatingComponent.prototype.ngOnInit = function () {
            this.max = typeof this.max !== 'undefined' ? this.max : 5;
            this.titles =
                typeof this.titles !== 'undefined' && this.titles.length > 0
                    ? this.titles
                    : [];
            this.range = this.buildTemplateObjects(this.max);
        };
        // model -> view
        RatingComponent.prototype.writeValue = function (value) {
            if (value % 1 !== value) {
                this.value = Math.round(value);
                this.preValue = value;
                this.changeDetection.markForCheck();
                return;
            }
            this.preValue = value;
            this.value = value;
            this.changeDetection.markForCheck();
        };
        RatingComponent.prototype.enter = function (value) {
            if (!this.readonly) {
                this.value = value;
                this.changeDetection.markForCheck();
                this.onHover.emit(value);
            }
        };
        RatingComponent.prototype.reset = function () {
            this.value = Math.round(this.preValue);
            this.changeDetection.markForCheck();
            this.onLeave.emit(this.value);
        };
        RatingComponent.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        RatingComponent.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        RatingComponent.prototype.rate = function (value) {
            if (!this.readonly && value >= 0 && value <= this.range.length) {
                this.writeValue(value);
                this.onChange(value);
            }
        };
        RatingComponent.prototype.buildTemplateObjects = function (max) {
            var result = [];
            for (var i = 0; i < max; i++) {
                result.push({
                    index: i,
                    title: this.titles[i] || i + 1
                });
            }
            return result;
        };
        /** @nocollapse */ RatingComponent.ɵfac = function RatingComponent_Factory(t) { return new (t || RatingComponent)(core["ɵɵdirectiveInject"](core.ChangeDetectorRef), core["ɵɵdirectiveInject"](RatingConfig)); };
        /** @nocollapse */ RatingComponent.ɵcmp = core["ɵɵdefineComponent"]({ type: RatingComponent, selectors: [["rating"]], hostBindings: function RatingComponent_HostBindings(rf, ctx) { if (rf & 1) {
                core["ɵɵlistener"]("keydown", function RatingComponent_keydown_HostBindingHandler($event) { return ctx.onKeydown($event); });
            } }, inputs: { max: "max", readonly: "readonly", titles: "titles", customTemplate: "customTemplate" }, outputs: { onHover: "onHover", onLeave: "onLeave" }, features: [core["ɵɵProvidersFeature"]([RATING_CONTROL_VALUE_ACCESSOR])], decls: 4, vars: 4, consts: [["tabindex", "0", "role", "slider", "aria-valuemin", "0", 3, "mouseleave", "keydown"], ["star", ""], ["ngFor", "", 3, "ngForOf"], [1, "sr-only"], [1, "bs-rating-star", 3, "title", "mouseenter", "click"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"]], template: function RatingComponent_Template(rf, ctx) { if (rf & 1) {
                core["ɵɵelementStart"](0, "span", 0);
                core["ɵɵlistener"]("mouseleave", function RatingComponent_Template_span_mouseleave_0_listener() { return ctx.reset(); })("keydown", function RatingComponent_Template_span_keydown_0_listener($event) { return ctx.onKeydown($event); });
                core["ɵɵtemplate"](1, RatingComponent_ng_template_1_Template, 1, 1, "ng-template", null, 1, core["ɵɵtemplateRefExtractor"]);
                core["ɵɵtemplate"](3, RatingComponent_ng_template_3_Template, 4, 11, "ng-template", 2);
                core["ɵɵelementEnd"]();
            } if (rf & 2) {
                core["ɵɵattribute"]("aria-label", ctx.ariaLabel)("aria-valuemax", ctx.range.length)("aria-valuenow", ctx.value);
                core["ɵɵadvance"](3);
                core["ɵɵproperty"]("ngForOf", ctx.range);
            } }, directives: [common.NgForOf, common.NgTemplateOutlet], encapsulation: 2, changeDetection: 0 });
        return RatingComponent;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](RatingComponent, [{
            type: core.Component,
            args: [{
                    selector: 'rating',
                    templateUrl: './rating.component.html',
                    providers: [RATING_CONTROL_VALUE_ACCESSOR],
                    changeDetection: core.ChangeDetectionStrategy.OnPush
                }]
        }], function () { return [{ type: core.ChangeDetectorRef }, { type: RatingConfig }]; }, { max: [{
                type: core.Input
            }], readonly: [{
                type: core.Input
            }], titles: [{
                type: core.Input
            }], customTemplate: [{
                type: core.Input
            }], onHover: [{
                type: core.Output
            }], onLeave: [{
                type: core.Output
            }], onKeydown: [{
                type: core.HostListener,
                args: ['keydown', ['$event']]
            }] }); })();

    var RatingModule = /** @class */ (function () {
        function RatingModule() {
        }
        RatingModule.forRoot = function () {
            return {
                ngModule: RatingModule,
                providers: [RatingConfig]
            };
        };
        /** @nocollapse */ RatingModule.ɵmod = core["ɵɵdefineNgModule"]({ type: RatingModule });
        /** @nocollapse */ RatingModule.ɵinj = core["ɵɵdefineInjector"]({ factory: function RatingModule_Factory(t) { return new (t || RatingModule)(); }, imports: [[common.CommonModule]] });
        return RatingModule;
    }());
    (function () { (typeof ngJitMode === "undefined" || ngJitMode) && core["ɵɵsetNgModuleScope"](RatingModule, { declarations: [RatingComponent], imports: [common.CommonModule], exports: [RatingComponent] }); })();
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](RatingModule, [{
            type: core.NgModule,
            args: [{
                    imports: [common.CommonModule],
                    declarations: [RatingComponent],
                    exports: [RatingComponent]
                }]
        }], null, null); })();

    exports.RatingComponent = RatingComponent;
    exports.RatingConfig = RatingConfig;
    exports.RatingModule = RatingModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
