// ngx-bootstrap/popover/esm5/popover.config.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Configuration service for the Popover directive.
 * You can inject this service, typically in your root component, and customize
 * the values of its properties in order to provide default values for all the
 * popovers used in the application.
 */
var PopoverConfig = /** @class */ (function () {
    function PopoverConfig() {
        /** sets disable adaptive position */
        this.adaptivePosition = true;
        /**
         * Placement of a popover. Accepts: "top", "bottom", "left", "right", "auto"
         */
        this.placement = 'top';
        /**
         * Specifies events that should trigger. Supports a space separated list of
         * event names.
         */
        this.triggers = 'click';
        this.outsideClick = false;
        /** delay before showing the tooltip */
        this.delay = 0;
    }
    /** @nocollapse */ PopoverConfig.ɵfac = function PopoverConfig_Factory(t) { return new (t || PopoverConfig)(); };
    /** @nocollapse */ PopoverConfig.ɵprov = i0.ɵɵdefineInjectable({ token: PopoverConfig, factory: PopoverConfig.ɵfac });
    return PopoverConfig;
}());
export { PopoverConfig };
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(PopoverConfig, [{
        type: Injectable
    }], null, null); })();
