// ngx-bootstrap/datepicker/esm2015/base/bs-datepicker-container.js
export class BsDatepickerAbstractComponent {
    constructor() {
        this.customRanges = [];
        this.chosenRange = [];
    }
    set minDate(value) {
        this._effects.setMinDate(value);
    }
    set maxDate(value) {
        this._effects.setMaxDate(value);
    }
    set daysDisabled(value) {
        this._effects.setDaysDisabled(value);
    }
    set datesDisabled(value) {
        this._effects.setDatesDisabled(value);
    }
    set datesEnabled(value) {
        this._effects.setDatesEnabled(value);
    }
    set isDisabled(value) {
        this._effects.setDisabled(value);
    }
    set dateCustomClasses(value) {
        this._effects.setDateCustomClasses(value);
    }
    setViewMode(event) { }
    navigateTo(event) { }
    dayHoverHandler(event) { }
    weekHoverHandler(event) { }
    monthHoverHandler(event) { }
    yearHoverHandler(event) { }
    daySelectHandler(day) { }
    monthSelectHandler(event) { }
    yearSelectHandler(event) { }
    setRangeOnCalendar(dates) { }
    setToday() { }
    /* tslint:disable-next-line: no-any */
    _stopPropagation(event) {
        event.stopPropagation();
    }
}
