// ngx-bootstrap/modal/esm5/modal-options.class.js
import { Injectable, InjectionToken } from '@angular/core';
import * as i0 from "@angular/core";
var ModalOptions = /** @class */ (function () {
    function ModalOptions() {
    }
    /** @nocollapse */ ModalOptions.ɵfac = function ModalOptions_Factory(t) { return new (t || ModalOptions)(); };
    /** @nocollapse */ ModalOptions.ɵprov = i0.ɵɵdefineInjectable({ token: ModalOptions, factory: ModalOptions.ɵfac });
    return ModalOptions;
}());
export { ModalOptions };
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(ModalOptions, [{
        type: Injectable
    }], null, null); })();
export var modalConfigDefaults = {
    backdrop: true,
    keyboard: true,
    focus: true,
    show: false,
    ignoreBackdropClick: false,
    class: '',
    animated: true,
    initialState: {}
};
export var MODAL_CONFIG_DEFAULT_OVERRIDE = new InjectionToken('override-default-config');
export var CLASS_NAME = {
    SCROLLBAR_MEASURER: 'modal-scrollbar-measure',
    BACKDROP: 'modal-backdrop',
    OPEN: 'modal-open',
    FADE: 'fade',
    IN: 'in',
    SHOW: 'show' // bs4
};
export var SELECTOR = {
    DIALOG: '.modal-dialog',
    DATA_TOGGLE: '[data-toggle="modal"]',
    DATA_DISMISS: '[data-dismiss="modal"]',
    FIXED_CONTENT: '.navbar-fixed-top, .navbar-fixed-bottom, .is-fixed'
};
export var TRANSITION_DURATIONS = {
    MODAL: 300,
    BACKDROP: 150
};
export var DISMISS_REASONS = {
    BACKRDOP: 'backdrop-click',
    ESC: 'esc'
};
