// ngx-bootstrap/carousel/esm5/carousel.config.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var CarouselConfig = /** @class */ (function () {
    function CarouselConfig() {
        /* Default interval of auto changing of slides */
        this.interval = 5000;
        /* Is loop of auto changing of slides can be paused */
        this.noPause = false;
        /* Is slides can wrap from the last to the first slide */
        this.noWrap = false;
        /* Show carousel-indicators */
        this.showIndicators = true;
        /* Slides can be paused on focus */
        this.pauseOnFocus = false;
        /* If `true` - carousel indicators indicate slides chunks works ONLY if singleSlideOffset = FALSE */
        this.indicatorsByChunk = false;
        /* If value more then 1 — carousel works in multilist mode */
        this.itemsPerSlide = 1;
        /* If `true` — carousel shifts by one element. By default carousel shifts by number
          of visible elements (itemsPerSlide field) */
        this.singleSlideOffset = false;
    }
    /** @nocollapse */ CarouselConfig.ɵfac = function CarouselConfig_Factory(t) { return new (t || CarouselConfig)(); };
    /** @nocollapse */ CarouselConfig.ɵprov = i0.ɵɵdefineInjectable({ token: CarouselConfig, factory: CarouselConfig.ɵfac });
    return CarouselConfig;
}());
export { CarouselConfig };
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(CarouselConfig, [{
        type: Injectable
    }], null, null); })();
