// ngx-bootstrap/tabs/bundles/ngx-bootstrap-tabs.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-bootstrap/tabs', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = global || self, factory((global['ngx-bootstrap'] = global['ngx-bootstrap'] || {}, global['ngx-bootstrap'].tabs = {}), global.ng.core, global.ng.common));
}(this, (function (exports, core, common) { 'use strict';

    var NgTranscludeDirective = /** @class */ (function () {
        function NgTranscludeDirective(viewRef) {
            this.viewRef = viewRef;
        }
        Object.defineProperty(NgTranscludeDirective.prototype, "ngTransclude", {
            /* tslint:disable-next-line:no-any */
            get: function () {
                return this._ngTransclude;
            },
            set: function (templateRef) {
                this._ngTransclude = templateRef;
                if (templateRef) {
                    this.viewRef.createEmbeddedView(templateRef);
                }
            },
            enumerable: true,
            configurable: true
        });
        /** @nocollapse */ NgTranscludeDirective.ɵfac = function NgTranscludeDirective_Factory(t) { return new (t || NgTranscludeDirective)(core["ɵɵdirectiveInject"](core.ViewContainerRef)); };
        /** @nocollapse */ NgTranscludeDirective.ɵdir = core["ɵɵdefineDirective"]({ type: NgTranscludeDirective, selectors: [["", "ngTransclude", ""]], inputs: { ngTransclude: "ngTransclude" } });
        return NgTranscludeDirective;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](NgTranscludeDirective, [{
            type: core.Directive,
            args: [{
                    selector: '[ngTransclude]'
                }]
        }], function () { return [{ type: core.ViewContainerRef }]; }, { ngTransclude: [{
                type: core.Input
            }] }); })();

    var TabsetConfig = /** @class */ (function () {
        function TabsetConfig() {
            /** provides default navigation context class: 'tabs' or 'pills' */
            this.type = 'tabs';
            /** provides possibility to set keyNavigations enable or disable, by default is enable */
            this.isKeysAllowed = true;
            /** aria label for tab list */
            this.ariaLabel = 'Tabs';
        }
        /** @nocollapse */ TabsetConfig.ɵfac = function TabsetConfig_Factory(t) { return new (t || TabsetConfig)(); };
        /** @nocollapse */ TabsetConfig.ɵprov = core["ɵɵdefineInjectable"]({ token: TabsetConfig, factory: TabsetConfig.ɵfac });
        return TabsetConfig;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](TabsetConfig, [{
            type: core.Injectable
        }], null, null); })();

    function TabsetComponent_li_1_span_4_Template(rf, ctx) { if (rf & 1) {
        var _r15 = core["ɵɵgetCurrentView"]();
        core["ɵɵelementStart"](0, "span", 7);
        core["ɵɵlistener"]("click", function TabsetComponent_li_1_span_4_Template_span_click_0_listener($event) { core["ɵɵrestoreView"](_r15); var tabz_r11 = core["ɵɵnextContext"]().$implicit; var ctx_r14 = core["ɵɵnextContext"](); $event.preventDefault(); return ctx_r14.removeTab(tabz_r11); });
        core["ɵɵtext"](1, " \u274C");
        core["ɵɵelementEnd"]();
    } }
    var _c0 = function (a1) { return ["nav-item", a1]; };
    function TabsetComponent_li_1_Template(rf, ctx) { if (rf & 1) {
        var _r18 = core["ɵɵgetCurrentView"]();
        core["ɵɵelementStart"](0, "li", 3);
        core["ɵɵlistener"]("keydown", function TabsetComponent_li_1_Template_li_keydown_0_listener($event) { core["ɵɵrestoreView"](_r18); var i_r12 = ctx.index; var ctx_r17 = core["ɵɵnextContext"](); return ctx_r17.keyNavActions($event, i_r12); });
        core["ɵɵelementStart"](1, "a", 4);
        core["ɵɵlistener"]("click", function TabsetComponent_li_1_Template_a_click_1_listener() { var tabz_r11 = ctx.$implicit; return tabz_r11.active = true; });
        core["ɵɵelementStart"](2, "span", 5);
        core["ɵɵtext"](3);
        core["ɵɵelementEnd"]();
        core["ɵɵtemplate"](4, TabsetComponent_li_1_span_4_Template, 2, 0, "span", 6);
        core["ɵɵelementEnd"]();
        core["ɵɵelementEnd"]();
    } if (rf & 2) {
        var tabz_r11 = ctx.$implicit;
        core["ɵɵclassProp"]("active", tabz_r11.active)("disabled", tabz_r11.disabled);
        core["ɵɵproperty"]("ngClass", core["ɵɵpureFunction1"](15, _c0, tabz_r11.customClass || ""));
        core["ɵɵadvance"](1);
        core["ɵɵclassProp"]("active", tabz_r11.active)("disabled", tabz_r11.disabled);
        core["ɵɵattribute"]("aria-controls", tabz_r11.id ? tabz_r11.id : "")("aria-selected", !!tabz_r11.active)("id", tabz_r11.id ? tabz_r11.id + "-link" : "");
        core["ɵɵadvance"](1);
        core["ɵɵproperty"]("ngTransclude", tabz_r11.headingRef);
        core["ɵɵadvance"](1);
        core["ɵɵtextInterpolate"](tabz_r11.heading);
        core["ɵɵadvance"](1);
        core["ɵɵproperty"]("ngIf", tabz_r11.removable);
    } }
    var _c1 = ["*"];
    // todo: add active event to tab
    // todo: fix? mixing static and dynamic tabs position tabs in order of creation
    var TabsetComponent = /** @class */ (function () {
        function TabsetComponent(config, renderer, elementRef) {
            this.renderer = renderer;
            this.elementRef = elementRef;
            this.clazz = true;
            this.tabs = [];
            this.classMap = {};
            Object.assign(this, config);
        }
        Object.defineProperty(TabsetComponent.prototype, "vertical", {
            /** if true tabs will be placed vertically */
            get: function () {
                return this._vertical;
            },
            set: function (value) {
                this._vertical = value;
                this.setClassMap();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TabsetComponent.prototype, "justified", {
            /** if true tabs fill the container and have a consistent width */
            get: function () {
                return this._justified;
            },
            set: function (value) {
                this._justified = value;
                this.setClassMap();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TabsetComponent.prototype, "type", {
            /** navigation context class: 'tabs' or 'pills' */
            get: function () {
                return this._type;
            },
            set: function (value) {
                this._type = value;
                this.setClassMap();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TabsetComponent.prototype, "isKeysAllowed", {
            get: function () {
                return this._isKeysAllowed;
            },
            set: function (value) {
                this._isKeysAllowed = value;
            },
            enumerable: true,
            configurable: true
        });
        TabsetComponent.prototype.ngOnDestroy = function () {
            this.isDestroyed = true;
        };
        TabsetComponent.prototype.addTab = function (tab) {
            this.tabs.push(tab);
            tab.active = this.tabs.length === 1 && typeof tab.active === 'undefined';
        };
        TabsetComponent.prototype.removeTab = function (tab, options) {
            if (options === void 0) { options = { reselect: true, emit: true }; }
            var index = this.tabs.indexOf(tab);
            if (index === -1 || this.isDestroyed) {
                return;
            }
            // Select a new tab if the tab to be removed is selected and not destroyed
            if (options.reselect && tab.active && this.hasAvailableTabs(index)) {
                var newActiveIndex = this.getClosestTabIndex(index);
                this.tabs[newActiveIndex].active = true;
            }
            if (options.emit) {
                tab.removed.emit(tab);
            }
            this.tabs.splice(index, 1);
            if (tab.elementRef.nativeElement.parentNode) {
                this.renderer.removeChild(tab.elementRef.nativeElement.parentNode, tab.elementRef.nativeElement);
            }
        };
        /* tslint:disable-next-line: cyclomatic-complexity */
        TabsetComponent.prototype.keyNavActions = function (event, index) {
            if (!this.isKeysAllowed) {
                return;
            }
            var list = Array.from(this.elementRef.nativeElement.querySelectorAll('.nav-link'));
            // const activeElList = list.filter((el: HTMLElement) => !el.classList.contains('disabled'));
            // tslint:disable-next-line:deprecation
            if (event.keyCode === 13 || event.key === 'Enter' || event.keyCode === 32 || event.key === 'Space') {
                event.preventDefault();
                var currentTab = list[(index) % list.length];
                currentTab.click();
                return;
            }
            // tslint:disable-next-line:deprecation
            if (event.keyCode === 39 || event.key === 'RightArrow') {
                var nextTab = void 0;
                var shift = 1;
                do {
                    nextTab = list[(index + shift) % list.length];
                    shift++;
                } while (nextTab.classList.contains('disabled'));
                nextTab.focus();
                return;
            }
            // tslint:disable-next-line:deprecation
            if (event.keyCode === 37 || event.key === 'LeftArrow') {
                var previousTab = void 0;
                var shift = 1;
                var i = index;
                do {
                    if ((i - shift) < 0) {
                        i = list.length - 1;
                        previousTab = list[i];
                        shift = 0;
                    }
                    else {
                        previousTab = list[i - shift];
                    }
                    shift++;
                } while (previousTab.classList.contains('disabled'));
                previousTab.focus();
                return;
            }
            // tslint:disable-next-line:deprecation
            if (event.keyCode === 36 || event.key === 'Home') {
                event.preventDefault();
                var firstTab = void 0;
                var shift = 0;
                do {
                    firstTab = list[shift % list.length];
                    shift++;
                } while (firstTab.classList.contains('disabled'));
                firstTab.focus();
                return;
            }
            // tslint:disable-next-line:deprecation
            if (event.keyCode === 35 || event.key === 'End') {
                event.preventDefault();
                var lastTab = void 0;
                var shift = 1;
                var i = index;
                do {
                    if ((i - shift) < 0) {
                        i = list.length - 1;
                        lastTab = list[i];
                        shift = 0;
                    }
                    else {
                        lastTab = list[i - shift];
                    }
                    shift++;
                } while (lastTab.classList.contains('disabled'));
                lastTab.focus();
                return;
            }
            // tslint:disable-next-line:deprecation
            if (event.keyCode === 46 || event.key === 'Delete') {
                if (this.tabs[index].removable) {
                    this.removeTab(this.tabs[index]);
                    if (list[index + 1]) {
                        list[(index + 1) % list.length].focus();
                        return;
                    }
                    if (list[list.length - 1]) {
                        list[0].focus();
                    }
                }
            }
        };
        TabsetComponent.prototype.getClosestTabIndex = function (index) {
            var tabsLength = this.tabs.length;
            if (!tabsLength) {
                return -1;
            }
            for (var step = 1; step <= tabsLength; step += 1) {
                var prevIndex = index - step;
                var nextIndex = index + step;
                if (this.tabs[prevIndex] && !this.tabs[prevIndex].disabled) {
                    return prevIndex;
                }
                if (this.tabs[nextIndex] && !this.tabs[nextIndex].disabled) {
                    return nextIndex;
                }
            }
            return -1;
        };
        TabsetComponent.prototype.hasAvailableTabs = function (index) {
            var tabsLength = this.tabs.length;
            if (!tabsLength) {
                return false;
            }
            for (var i = 0; i < tabsLength; i += 1) {
                if (!this.tabs[i].disabled && i !== index) {
                    return true;
                }
            }
            return false;
        };
        TabsetComponent.prototype.setClassMap = function () {
            var _a;
            this.classMap = (_a = {
                    'nav-stacked': this.vertical,
                    'flex-column': this.vertical,
                    'nav-justified': this.justified
                },
                _a["nav-" + this.type] = true,
                _a);
        };
        /** @nocollapse */ TabsetComponent.ɵfac = function TabsetComponent_Factory(t) { return new (t || TabsetComponent)(core["ɵɵdirectiveInject"](TabsetConfig), core["ɵɵdirectiveInject"](core.Renderer2), core["ɵɵdirectiveInject"](core.ElementRef)); };
        /** @nocollapse */ TabsetComponent.ɵcmp = core["ɵɵdefineComponent"]({ type: TabsetComponent, selectors: [["tabset"]], hostVars: 2, hostBindings: function TabsetComponent_HostBindings(rf, ctx) { if (rf & 2) {
                core["ɵɵclassProp"]("tab-container", ctx.clazz);
            } }, inputs: { vertical: "vertical", justified: "justified", type: "type" }, ngContentSelectors: _c1, decls: 4, vars: 3, consts: [["role", "tablist", 1, "nav", 3, "ngClass", "click"], [3, "ngClass", "active", "disabled", "keydown", 4, "ngFor", "ngForOf"], [1, "tab-content"], [3, "ngClass", "keydown"], ["href", "javascript:void(0);", "role", "tab", 1, "nav-link", 3, "click"], [3, "ngTransclude"], ["class", "bs-remove-tab", 3, "click", 4, "ngIf"], [1, "bs-remove-tab", 3, "click"]], template: function TabsetComponent_Template(rf, ctx) { if (rf & 1) {
                core["ɵɵprojectionDef"]();
                core["ɵɵelementStart"](0, "ul", 0);
                core["ɵɵlistener"]("click", function TabsetComponent_Template_ul_click_0_listener($event) { return $event.preventDefault(); });
                core["ɵɵtemplate"](1, TabsetComponent_li_1_Template, 5, 17, "li", 1);
                core["ɵɵelementEnd"]();
                core["ɵɵelementStart"](2, "div", 2);
                core["ɵɵprojection"](3);
                core["ɵɵelementEnd"]();
            } if (rf & 2) {
                core["ɵɵproperty"]("ngClass", ctx.classMap);
                core["ɵɵattribute"]("aria-label", ctx.ariaLabel);
                core["ɵɵadvance"](1);
                core["ɵɵproperty"]("ngForOf", ctx.tabs);
            } }, directives: [common.NgClass, common.NgForOf, NgTranscludeDirective, common.NgIf], styles: ["[_nghost-%COMP%]   .nav-tabs[_ngcontent-%COMP%]   .nav-item.disabled[_ngcontent-%COMP%]   a.disabled[_ngcontent-%COMP%]{cursor:default}"] });
        return TabsetComponent;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](TabsetComponent, [{
            type: core.Component,
            args: [{
                    selector: 'tabset',
                    templateUrl: './tabset.component.html',
                    styleUrls: ['./tabs.scss']
                }]
        }], function () { return [{ type: TabsetConfig }, { type: core.Renderer2 }, { type: core.ElementRef }]; }, { vertical: [{
                type: core.Input
            }], justified: [{
                type: core.Input
            }], type: [{
                type: core.Input
            }], clazz: [{
                type: core.HostBinding,
                args: ['class.tab-container']
            }] }); })();

    var TabDirective = /** @class */ (function () {
        function TabDirective(tabset, elementRef, renderer) {
            this.elementRef = elementRef;
            this.renderer = renderer;
            /** fired when tab became active, $event:Tab equals to selected instance of Tab component */
            this.selectTab = new core.EventEmitter();
            /** fired when tab became inactive, $event:Tab equals to deselected instance of Tab component */
            this.deselect = new core.EventEmitter();
            /** fired before tab will be removed, $event:Tab equals to instance of removed tab */
            this.removed = new core.EventEmitter();
            this.addClass = true;
            this.role = 'tabpanel';
            this.tabset = tabset;
            this.tabset.addTab(this);
        }
        Object.defineProperty(TabDirective.prototype, "customClass", {
            /** if set, will be added to the tab's class attribute. Multiple classes are supported. */
            get: function () {
                return this._customClass;
            },
            set: function (customClass) {
                var _this = this;
                if (this.customClass) {
                    this.customClass.split(' ').forEach(function (cssClass) {
                        _this.renderer.removeClass(_this.elementRef.nativeElement, cssClass);
                    });
                }
                this._customClass = customClass ? customClass.trim() : null;
                if (this.customClass) {
                    this.customClass.split(' ').forEach(function (cssClass) {
                        _this.renderer.addClass(_this.elementRef.nativeElement, cssClass);
                    });
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TabDirective.prototype, "active", {
            /** tab active state toggle */
            get: function () {
                return this._active;
            },
            set: function (active) {
                var _this = this;
                if (this._active === active) {
                    return;
                }
                if ((this.disabled && active) || !active) {
                    if (this._active && !active) {
                        this.deselect.emit(this);
                        this._active = active;
                    }
                    return;
                }
                this._active = active;
                this.selectTab.emit(this);
                this.tabset.tabs.forEach(function (tab) {
                    if (tab !== _this) {
                        tab.active = false;
                    }
                });
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TabDirective.prototype, "ariaLabelledby", {
            get: function () {
                return this.id ? this.id + "-link" : '';
            },
            enumerable: true,
            configurable: true
        });
        TabDirective.prototype.ngOnInit = function () {
            this.removable = this.removable;
        };
        TabDirective.prototype.ngOnDestroy = function () {
            this.tabset.removeTab(this, { reselect: false, emit: false });
        };
        /** @nocollapse */ TabDirective.ɵfac = function TabDirective_Factory(t) { return new (t || TabDirective)(core["ɵɵdirectiveInject"](TabsetComponent), core["ɵɵdirectiveInject"](core.ElementRef), core["ɵɵdirectiveInject"](core.Renderer2)); };
        /** @nocollapse */ TabDirective.ɵdir = core["ɵɵdefineDirective"]({ type: TabDirective, selectors: [["tab"], ["", "tab", ""]], hostVars: 7, hostBindings: function TabDirective_HostBindings(rf, ctx) { if (rf & 2) {
                core["ɵɵattribute"]("id", ctx.id)("role", ctx.role)("aria-labelledby", ctx.ariaLabelledby);
                core["ɵɵclassProp"]("active", ctx.active)("tab-pane", ctx.addClass);
            } }, inputs: { heading: "heading", id: "id", disabled: "disabled", removable: "removable", customClass: "customClass", active: "active" }, outputs: { selectTab: "selectTab", deselect: "deselect", removed: "removed" } });
        return TabDirective;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](TabDirective, [{
            type: core.Directive,
            args: [{ selector: 'tab, [tab]' }]
        }], function () { return [{ type: TabsetComponent }, { type: core.ElementRef }, { type: core.Renderer2 }]; }, { heading: [{
                type: core.Input
            }], id: [{
                type: core.HostBinding,
                args: ['attr.id']
            }, {
                type: core.Input
            }], disabled: [{
                type: core.Input
            }], removable: [{
                type: core.Input
            }], customClass: [{
                type: core.Input
            }], active: [{
                type: core.HostBinding,
                args: ['class.active']
            }, {
                type: core.Input
            }], selectTab: [{
                type: core.Output
            }], deselect: [{
                type: core.Output
            }], removed: [{
                type: core.Output
            }], addClass: [{
                type: core.HostBinding,
                args: ['class.tab-pane']
            }], role: [{
                type: core.HostBinding,
                args: ['attr.role']
            }], ariaLabelledby: [{
                type: core.HostBinding,
                args: ['attr.aria-labelledby']
            }] }); })();

    /** Should be used to mark <ng-template> element as a template for tab heading */
    var TabHeadingDirective = /** @class */ (function () {
        /* tslint:disable-next-line:no-any */
        function TabHeadingDirective(templateRef, tab) {
            tab.headingRef = templateRef;
        }
        /** @nocollapse */ TabHeadingDirective.ɵfac = function TabHeadingDirective_Factory(t) { return new (t || TabHeadingDirective)(core["ɵɵdirectiveInject"](core.TemplateRef), core["ɵɵdirectiveInject"](TabDirective)); };
        /** @nocollapse */ TabHeadingDirective.ɵdir = core["ɵɵdefineDirective"]({ type: TabHeadingDirective, selectors: [["", "tabHeading", ""]] });
        return TabHeadingDirective;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](TabHeadingDirective, [{
            type: core.Directive,
            args: [{ selector: '[tabHeading]' }]
        }], function () { return [{ type: core.TemplateRef }, { type: TabDirective }]; }, null); })();

    var TabsModule = /** @class */ (function () {
        function TabsModule() {
        }
        TabsModule.forRoot = function () {
            return {
                ngModule: TabsModule,
                providers: [TabsetConfig]
            };
        };
        /** @nocollapse */ TabsModule.ɵmod = core["ɵɵdefineNgModule"]({ type: TabsModule });
        /** @nocollapse */ TabsModule.ɵinj = core["ɵɵdefineInjector"]({ factory: function TabsModule_Factory(t) { return new (t || TabsModule)(); }, imports: [[common.CommonModule]] });
        return TabsModule;
    }());
    (function () { (typeof ngJitMode === "undefined" || ngJitMode) && core["ɵɵsetNgModuleScope"](TabsModule, { declarations: [NgTranscludeDirective,
            TabDirective,
            TabsetComponent,
            TabHeadingDirective], imports: [common.CommonModule], exports: [TabDirective,
            TabsetComponent,
            TabHeadingDirective,
            NgTranscludeDirective] }); })();
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](TabsModule, [{
            type: core.NgModule,
            args: [{
                    imports: [common.CommonModule],
                    declarations: [
                        NgTranscludeDirective,
                        TabDirective,
                        TabsetComponent,
                        TabHeadingDirective
                    ],
                    exports: [
                        TabDirective,
                        TabsetComponent,
                        TabHeadingDirective,
                        NgTranscludeDirective
                    ]
                }]
        }], null, null); })();

    exports.NgTranscludeDirective = NgTranscludeDirective;
    exports.TabDirective = TabDirective;
    exports.TabHeadingDirective = TabHeadingDirective;
    exports.TabsModule = TabsModule;
    exports.TabsetComponent = TabsetComponent;
    exports.TabsetConfig = TabsetConfig;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
