// ngx-bootstrap/popover/bundles/ngx-bootstrap-popover.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('ngx-bootstrap/component-loader'), require('ngx-bootstrap/utils'), require('@angular/common'), require('ngx-bootstrap/positioning'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('ngx-bootstrap/popover', ['exports', '@angular/core', 'ngx-bootstrap/component-loader', 'ngx-bootstrap/utils', '@angular/common', 'ngx-bootstrap/positioning', 'rxjs'], factory) :
    (global = global || self, factory((global['ngx-bootstrap'] = global['ngx-bootstrap'] || {}, global['ngx-bootstrap'].popover = {}), global.ng.core, global.componentLoader, global.utils, global.ng.common, global.positioning, global.rxjs));
}(this, (function (exports, core, componentLoader, utils, common, positioning, rxjs) { 'use strict';

    /**
     * Configuration service for the Popover directive.
     * You can inject this service, typically in your root component, and customize
     * the values of its properties in order to provide default values for all the
     * popovers used in the application.
     */
    var PopoverConfig = /** @class */ (function () {
        function PopoverConfig() {
            /** sets disable adaptive position */
            this.adaptivePosition = true;
            /**
             * Placement of a popover. Accepts: "top", "bottom", "left", "right", "auto"
             */
            this.placement = 'top';
            /**
             * Specifies events that should trigger. Supports a space separated list of
             * event names.
             */
            this.triggers = 'click';
            this.outsideClick = false;
            /** delay before showing the tooltip */
            this.delay = 0;
        }
        /** @nocollapse */ PopoverConfig.ɵfac = function PopoverConfig_Factory(t) { return new (t || PopoverConfig)(); };
        /** @nocollapse */ PopoverConfig.ɵprov = core["ɵɵdefineInjectable"]({ token: PopoverConfig, factory: PopoverConfig.ɵfac });
        return PopoverConfig;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](PopoverConfig, [{
            type: core.Injectable
        }], null, null); })();

    function PopoverContainerComponent_h3_1_Template(rf, ctx) { if (rf & 1) {
        core["ɵɵelementStart"](0, "h3", 3);
        core["ɵɵtext"](1);
        core["ɵɵelementEnd"]();
    } if (rf & 2) {
        var ctx_r1 = core["ɵɵnextContext"]();
        core["ɵɵadvance"](1);
        core["ɵɵtextInterpolate"](ctx_r1.title);
    } }
    var _c0 = ["*"];
    var PopoverContainerComponent = /** @class */ (function () {
        function PopoverContainerComponent(config) {
            Object.assign(this, config);
        }
        Object.defineProperty(PopoverContainerComponent.prototype, "isBs3", {
            get: function () {
                return utils.isBs3();
            },
            enumerable: true,
            configurable: true
        });
        /** @nocollapse */ PopoverContainerComponent.ɵfac = function PopoverContainerComponent_Factory(t) { return new (t || PopoverContainerComponent)(core["ɵɵdirectiveInject"](PopoverConfig)); };
        /** @nocollapse */ PopoverContainerComponent.ɵcmp = core["ɵɵdefineComponent"]({ type: PopoverContainerComponent, selectors: [["popover-container"]], hostAttrs: ["role", "tooltip", 2, "display", "block"], hostVars: 6, hostBindings: function PopoverContainerComponent_HostBindings(rf, ctx) { if (rf & 2) {
                core["ɵɵclassMap"]("popover in popover-" + ctx.placement + " " + "bs-popover-" + ctx.placement + " " + ctx.placement + " " + ctx.containerClass);
                core["ɵɵclassProp"]("show", !ctx.isBs3)("bs3", ctx.isBs3);
            } }, inputs: { placement: "placement", title: "title" }, ngContentSelectors: _c0, decls: 4, vars: 1, consts: [[1, "popover-arrow", "arrow"], ["class", "popover-title popover-header", 4, "ngIf"], [1, "popover-content", "popover-body"], [1, "popover-title", "popover-header"]], template: function PopoverContainerComponent_Template(rf, ctx) { if (rf & 1) {
                core["ɵɵprojectionDef"]();
                core["ɵɵelement"](0, "div", 0);
                core["ɵɵtemplate"](1, PopoverContainerComponent_h3_1_Template, 2, 1, "h3", 1);
                core["ɵɵelementStart"](2, "div", 2);
                core["ɵɵprojection"](3);
                core["ɵɵelementEnd"]();
            } if (rf & 2) {
                core["ɵɵadvance"](1);
                core["ɵɵproperty"]("ngIf", ctx.title);
            } }, directives: [common.NgIf], styles: [".bs3.popover-top[_nghost-%COMP%] {\n      margin-bottom: 10px;\n    }\n    .bs3.popover.top[_nghost-%COMP%] > .arrow[_ngcontent-%COMP%] {\n      margin-left: -2px;\n    }\n    .bs3.popover.top[_nghost-%COMP%] {\n      margin-bottom: 10px;\n    }\n    .popover.bottom[_nghost-%COMP%] > .arrow[_ngcontent-%COMP%] {\n      margin-left: -4px;\n    }\n    .bs3.bs-popover-left[_nghost-%COMP%] {\n      margin-right: .5rem;\n    }\n    .bs3.bs-popover-right[_nghost-%COMP%]   .arrow[_ngcontent-%COMP%], .bs3.bs-popover-left[_nghost-%COMP%]   .arrow[_ngcontent-%COMP%]{\n      margin: .3rem 0;\n    }"], changeDetection: 0 });
        return PopoverContainerComponent;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](PopoverContainerComponent, [{
            type: core.Component,
            args: [{
                    selector: 'popover-container',
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                    // tslint:disable-next-line
                    host: {
                        '[class]': '"popover in popover-" + placement + " " + "bs-popover-" + placement + " " + placement + " " + containerClass',
                        '[class.show]': '!isBs3',
                        '[class.bs3]': 'isBs3',
                        role: 'tooltip',
                        style: 'display:block;'
                    },
                    styles: [
                        "\n    :host.bs3.popover-top {\n      margin-bottom: 10px;\n    }\n    :host.bs3.popover.top>.arrow {\n      margin-left: -2px;\n    }\n    :host.bs3.popover.top {\n      margin-bottom: 10px;\n    }\n    :host.popover.bottom>.arrow {\n      margin-left: -4px;\n    }\n    :host.bs3.bs-popover-left {\n      margin-right: .5rem;\n    }\n    :host.bs3.bs-popover-right .arrow, :host.bs3.bs-popover-left .arrow{\n      margin: .3rem 0;\n    }\n    "
                    ],
                    templateUrl: './popover-container.component.html'
                }]
        }], function () { return [{ type: PopoverConfig }]; }, { placement: [{
                type: core.Input
            }], title: [{
                type: core.Input
            }] }); })();

    /**
     * A lightweight, extensible directive for fancy popover creation.
     */
    var PopoverDirective = /** @class */ (function () {
        function PopoverDirective(_config, _elementRef, _renderer, _viewContainerRef, cis, _positionService) {
            this._elementRef = _elementRef;
            this._renderer = _renderer;
            this._positionService = _positionService;
            /**
             * Close popover on outside click
             */
            this.outsideClick = false;
            /**
             * Css class for popover container
             */
            this.containerClass = '';
            this._isInited = false;
            this._popover = cis
                .createLoader(_elementRef, _viewContainerRef, _renderer)
                .provide({ provide: PopoverConfig, useValue: _config });
            Object.assign(this, _config);
            this.onShown = this._popover.onShown;
            this.onHidden = this._popover.onHidden;
            // fix: no focus on button on Mac OS #1795
            if (typeof window !== 'undefined') {
                _elementRef.nativeElement.addEventListener('click', function () {
                    try {
                        _elementRef.nativeElement.focus();
                    }
                    catch (err) {
                        return;
                    }
                });
            }
        }
        Object.defineProperty(PopoverDirective.prototype, "isOpen", {
            /**
             * Returns whether or not the popover is currently being shown
             */
            get: function () {
                return this._popover.isShown;
            },
            set: function (value) {
                if (value) {
                    this.show();
                }
                else {
                    this.hide();
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Opens an element’s popover. This is considered a “manual” triggering of
         * the popover.
         */
        PopoverDirective.prototype.show = function () {
            var _this = this;
            if (this._popover.isShown || !this.popover || this._delayTimeoutId) {
                return;
            }
            this._positionService.setOptions({
                modifiers: {
                    flip: {
                        enabled: this.adaptivePosition
                    },
                    preventOverflow: {
                        enabled: this.adaptivePosition
                    }
                }
            });
            var showPopover = function () {
                if (_this._delayTimeoutId) {
                    _this._delayTimeoutId = undefined;
                }
                _this._popover
                    .attach(PopoverContainerComponent)
                    .to(_this.container)
                    .position({ attachment: _this.placement })
                    .show({
                    content: _this.popover,
                    context: _this.popoverContext,
                    placement: _this.placement,
                    title: _this.popoverTitle,
                    containerClass: _this.containerClass
                });
                if (!_this.adaptivePosition) {
                    _this._positionService.calcPosition();
                    _this._positionService.deletePositionElement(_this._popover._componentRef.location);
                }
                _this.isOpen = true;
            };
            var cancelDelayedTooltipShowing = function () {
                if (_this._popoverCancelShowFn) {
                    _this._popoverCancelShowFn();
                }
            };
            if (this.delay) {
                var _timer_1 = rxjs.timer(this.delay).subscribe(function () {
                    showPopover();
                    cancelDelayedTooltipShowing();
                });
                if (this.triggers) {
                    utils.parseTriggers(this.triggers)
                        .forEach(function (trigger) {
                        _this._popoverCancelShowFn = _this._renderer.listen(_this._elementRef.nativeElement, trigger.close, function () {
                            _timer_1.unsubscribe();
                            cancelDelayedTooltipShowing();
                        });
                    });
                }
            }
            else {
                showPopover();
            }
        };
        /**
         * Closes an element’s popover. This is considered a “manual” triggering of
         * the popover.
         */
        PopoverDirective.prototype.hide = function () {
            if (this._delayTimeoutId) {
                clearTimeout(this._delayTimeoutId);
                this._delayTimeoutId = undefined;
            }
            if (this.isOpen) {
                this._popover.hide();
                this.isOpen = false;
            }
        };
        /**
         * Toggles an element’s popover. This is considered a “manual” triggering of
         * the popover.
         */
        PopoverDirective.prototype.toggle = function () {
            if (this.isOpen) {
                return this.hide();
            }
            this.show();
        };
        PopoverDirective.prototype.ngOnInit = function () {
            var _this = this;
            // fix: seems there are an issue with `routerLinkActive`
            // which result in duplicated call ngOnInit without call to ngOnDestroy
            // read more: https://github.com/valor-software/ngx-bootstrap/issues/1885
            if (this._isInited) {
                return;
            }
            this._isInited = true;
            this._popover.listen({
                triggers: this.triggers,
                outsideClick: this.outsideClick,
                show: function () { return _this.show(); }
            });
        };
        PopoverDirective.prototype.ngOnDestroy = function () {
            this._popover.dispose();
        };
        /** @nocollapse */ PopoverDirective.ɵfac = function PopoverDirective_Factory(t) { return new (t || PopoverDirective)(core["ɵɵdirectiveInject"](PopoverConfig), core["ɵɵdirectiveInject"](core.ElementRef), core["ɵɵdirectiveInject"](core.Renderer2), core["ɵɵdirectiveInject"](core.ViewContainerRef), core["ɵɵdirectiveInject"](componentLoader.ComponentLoaderFactory), core["ɵɵdirectiveInject"](positioning.PositioningService)); };
        /** @nocollapse */ PopoverDirective.ɵdir = core["ɵɵdefineDirective"]({ type: PopoverDirective, selectors: [["", "popover", ""]], inputs: { adaptivePosition: "adaptivePosition", popover: "popover", popoverContext: "popoverContext", popoverTitle: "popoverTitle", placement: "placement", outsideClick: "outsideClick", triggers: "triggers", container: "container", containerClass: "containerClass", isOpen: "isOpen", delay: "delay" }, outputs: { onShown: "onShown", onHidden: "onHidden" }, exportAs: ["bs-popover"] });
        return PopoverDirective;
    }());
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](PopoverDirective, [{
            type: core.Directive,
            args: [{ selector: '[popover]', exportAs: 'bs-popover' }]
        }], function () { return [{ type: PopoverConfig }, { type: core.ElementRef }, { type: core.Renderer2 }, { type: core.ViewContainerRef }, { type: componentLoader.ComponentLoaderFactory }, { type: positioning.PositioningService }]; }, { adaptivePosition: [{
                type: core.Input
            }], popover: [{
                type: core.Input
            }], popoverContext: [{
                type: core.Input
            }], popoverTitle: [{
                type: core.Input
            }], placement: [{
                type: core.Input
            }], outsideClick: [{
                type: core.Input
            }], triggers: [{
                type: core.Input
            }], container: [{
                type: core.Input
            }], containerClass: [{
                type: core.Input
            }], isOpen: [{
                type: core.Input
            }], delay: [{
                type: core.Input
            }], onShown: [{
                type: core.Output
            }], onHidden: [{
                type: core.Output
            }] }); })();

    var PopoverModule = /** @class */ (function () {
        function PopoverModule() {
        }
        PopoverModule.forRoot = function () {
            return {
                ngModule: PopoverModule,
                providers: [PopoverConfig, componentLoader.ComponentLoaderFactory, positioning.PositioningService]
            };
        };
        /** @nocollapse */ PopoverModule.ɵmod = core["ɵɵdefineNgModule"]({ type: PopoverModule });
        /** @nocollapse */ PopoverModule.ɵinj = core["ɵɵdefineInjector"]({ factory: function PopoverModule_Factory(t) { return new (t || PopoverModule)(); }, imports: [[common.CommonModule]] });
        return PopoverModule;
    }());
    (function () { (typeof ngJitMode === "undefined" || ngJitMode) && core["ɵɵsetNgModuleScope"](PopoverModule, { declarations: [PopoverDirective, PopoverContainerComponent], imports: [common.CommonModule], exports: [PopoverDirective] }); })();
    /*@__PURE__*/ (function () { core["ɵsetClassMetadata"](PopoverModule, [{
            type: core.NgModule,
            args: [{
                    imports: [common.CommonModule],
                    declarations: [PopoverDirective, PopoverContainerComponent],
                    exports: [PopoverDirective],
                    entryComponents: [PopoverContainerComponent]
                }]
        }], null, null); })();

    exports.PopoverConfig = PopoverConfig;
    exports.PopoverContainerComponent = PopoverContainerComponent;
    exports.PopoverDirective = PopoverDirective;
    exports.PopoverModule = PopoverModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
